subsystems/drivesubsystem.py

```diff
@@ -281,9 +311,28 @@ def getSwerveAngle(self) -> Rotation2d:
         return Rotation2d(swerveAngle)
 
     def setSwerveAngle(self, swerveAngle: Rotation2d) -> None:
-        steerEncoderPulses = (
-            swerveAngle.radians() * constants.kSwerveEncoderPulsesPerRadian
+        currentAngle = self.getSwerveAngle().radians()
+
+        closestFullRotation = (
+            floor(abs(currentAngle / tau)) * (-1 if currentAngle < 0 else 1) * tau
         )
+
+        currentOptimalAngle = swerveAngle.radians() + closestFullRotation - currentAngle
+
+        potentialNewAngles = [
+            currentOptimalAngle,
+            currentOptimalAngle - tau,
+            currentOptimalAngle + tau,
+        ]  # closest other options
+
+        swerveError = tau  # max possible error, a full rotation!
+        for potentialAngle in potentialNewAngles:
+            if abs(swerveError) > abs(potentialAngle):
+                swerveError = potentialAngle
+
+        steerEncoderPulses = (
+            currentAngle + swerveError
+        ) * constants.kSwerveEncoderPulsesPerRadian
```

Comment:
What's the motivation for putting the algorithm here for CTRE only? I don't think this algorithm should be hardware-dependent.

Reply:
If the core issue is that Rotation2d normalizes internally to a single turn, that would point to Rotation2d being the wrong class to use to represent an input angle for the module.

Reply:
the issue is the return value of optimization of the state being clamped, rotation2d allows for outside of one rotation